File: prng-compare/mt-test.cpp
#include <vector>
#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>
#include <fstream>
#include <random>
#include <limits>
#include <sstream>
#include <iterator>
#include <set>

using namespace std;

unsigned int seed;

int main(int argc, char *argv[]) 
{
    // Get PRNG seed
    ifstream seedFile;
    seedFile.open("seedFile.txt");
    if (seedFile.fail()) {  // If no file exists, create one
        // Generate seed from datetime function
        seed = time(NULL);

        // Open a file to store seed
        ofstream seedFile;
        seedFile.open("seedFile.txt");

        // Write seed value to file
        string SeedString = to_string(seed);
        seedFile << SeedString;

        seedFile.close();  // Close seed file
    }
    else {  // If a file exists, set the seed to the stored seed
        string seedLine;
        getline(seedFile, seedLine);
        seed = stoi(seedLine);
    }

    
    // Initialize and seed MT generator
    mt19937 mt(seed);
    //mt.seed(seed);

    // Print state of MT generator
    //cout << mt;

    // Get state from state file
    int sameState = atoi(argv[2]);
    if (sameState == 1) {
        ifstream stateFile("stateFile.txt");
        stateFile >> mt;
    }

    //cout << mt << endl;  // Check state;
    

    // Write generator maximum to file
    ofstream randMaxFile("../../RAND_MAX.txt");
    randMaxFile << mt.max();


    int randGen = atoi(argv[1]);
    // Write 3 random integers and 3 random floats to output files
    if (randGen == 0) {
        ofstream outFile("out-cpp.txt");
        for (size_t i = 1; i <= 5; ++i) {
            outFile << mt() << endl;
        }
    }

    if (randGen == 1) {
        ofstream outFloatFile("out-cpp-float.txt");
        for (size_t i = 1; i <= 5; ++i) {
            uniform_real_distribution<float> float_dist(0, 1);
            outFloatFile << float_dist(mt) << endl;
            double rand2 = float_dist(mt);
        }
    }

    if (randGen == 2) {
        ofstream outUniformFile("out-cpp-uni");
        for (size_t i = 1; i <=5; ++i) {
            uniform_int_distribution<u_int32_t> dist(0, mt.max());
            outUniformFile << dist(mt) << endl;
        }
    }




    // Write 3 random numbers from uniform int distribution to output file
    /*uniform_int_distribution<> uni_dist(0, mt.max());
    ofstream outUni("out-uniform-cpp.txt");
    for (size_t i = 1; i <= 3; ++i) {
        size_t rand_uni = uni_dist(mt);
        outUni << rand_uni << endl;
    }*/
}

File: prng-compare/mt-test.py
import numpy as np
from numpy.random import MT19937, SeedSequence
import os
import random
import datetime
from datetime import datetime
from pandas import UInt32Dtype
import sys
import os

# Set random number seed and write to file
if (sys.argv[1] != "null"):
    seed = int(sys.argv[1])
    seed_file = open("seedFile.txt", "w")
    seed_file.write(str(seed))
elif os.path.isfile("seedFile.txt"):
    # If seed file already exists, read from file
    seed_file = open("seedFile.txt", "r")
    seed = int(seed_file.read())
elif not os.path.isfile("seedFile.txt") or os.stat("seedFile.txt").st_size==0:
    # If no seed file exists, generate new seed from time method
    seed = int(datetime.utcnow().timestamp())

    # Create seed file and write to file
    seed_file = open("seedFile.txt", "w")
    seed_file.write(str(seed))

seed_file.close()  # Close seed file

# Set seed for numpy and random libraries
np.random.seed(seed)  # Set global numpy seed
random.seed(seed)




# Initialize and seed numpy mt generator
np.random.MT19937(seed)

if (int(sys.argv[2]) == 1):
    sameState = int(sys.argv[2])
else:
    sameState = 0

if (sameState == 1):
    if os.path.isfile("stateFile.txt"):
        stateFile = open("stateFile.txt")
        state_lines = stateFile.read().split(' ')
        state = [int(i) for i in state_lines[:-1]]
        np_state = ('MT19937', state, 624, 0, 0.0)
        np.random.set_state(np_state)
    else:
        state = [int(s) for s in list(np.random.get_state()[1])]
        np_state = ('MT19937', state, 624, 0, 0.0)
        # Write numpy mt state to file
        stateFile = open("stateFile.txt", 'w')
        for state_int in state:       
            stateFile.write(str(state_int) + " ")

    # Set random library state the same as numpy
    state.append(624)
    state = tuple(state)
    state = (3, tuple(state), None)
    random.setstate(state)


# Numpy output files
outNumpy = open("out-numpy.txt", 'w')
outNpFloat = open("out-np-float.txt", 'w')
#outNumpyUniform = open("out-np-uniform.txt", 'w')

# Random library output files
outFile = open("out-py.txt", 'w')
outFloat = open("out-py-float.txt", 'w')
outUniform = open("out-py-uni.txt", 'w')

# Get maximum random value
rand_max_file = open("../../RAND_MAX.txt")
RAND_MAX = int(rand_max_file.read())


for i in range(10):
    # Test random floats
    rand_float_np = np.random.rand()
    outNpFloat.write("{}\n".format(rand_float_np))

np.random.seed(seed) 
if (sameState == 1): 
    np.random.set_state(np_state)
for i in range(10):
    rand_uni_np = np.random.uniform(0,1)
    outUniform.write("{}\n".format(rand_uni_np))

random.seed(seed) 
if (sameState == 1): random.setstate(state)
for i in range(10):
    rand_float = random.random()
    outFloat.write("{}\n".format(rand_float))



# Set seed for numpy and random libraries
np.random.seed(seed) 
if (sameState == 1): 
    np.random.set_state(np_state)
for i in range(10):
    # Test randint
    #rand_int_numpy = np.random.randint(256**4, dtype='<u4')
    rand_int_numpy = np.random.randint(256**4)
    outNumpy.write("{}\n".format(rand_int_numpy))

random.seed(seed) 
if (sameState == 1): random.setstate(state)
for i in range(10):
    rand_int = random.randint(0, 256**4-1)
    outFile.write("{}\n".format(rand_int))


